extension/k8sleaderelector/extension.go

```diff
@@ -40,21 +58,47 @@ type leaderElectionExtension struct {
 	cancel        context.CancelFunc
 	waitGroup     sync.WaitGroup
 
-	onStartedLeading []StartCallback
-	onStoppedLeading []StopCallback
+	callBackChan  chan callBackFuncs
+	callBackFuncs []callBackFuncs
+
+	mu sync.Mutex
 }
 
 // If the receiver sets a callback function then it would be invoked when the leader wins the election
 func (lee *leaderElectionExtension) startedLeading(ctx context.Context) {
-	for _, callback := range lee.onStartedLeading {
-		callback(ctx)
+	// Create a channel for receivers which have registered the callback after the extension has become the leader. In such case the callback is pushed to
+	// channel and executed immediately.
+	lee.callBackChan = make(chan callBackFuncs, 1)
+
+	for _, callback := range lee.callBackFuncs {
+		callback.onStartLeading(ctx)
 	}
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				callback, ok := <-lee.callBackChan
```

Comment:
why do you have `ok` here? It make the busy loop which is pretty costly. You need to block when channel is emty